### ISSUE TITLE
CI failure: `pip freeze` no longer protects `setuptools`, `distribute` or `wheel` on Python 3.12+

### ISSUE THREAD

**Issue by ncoghlan** [jazzband/pip-tools#2104]
Some CI tests are failing due to this change in `pip` to no longer protect `setuptools`, `distribute` or `wheel` by default on Python 3.12+: https://github.com/pypa/pip/commit/5dc65eabb75f89d4f4749b6c764042c227f6870a

That has been in the `pip` dev branch for a while, and was published in `pip 24.1` in the past few days.

This change affects the `pip-tools` test suite by way of the `PACKAGES_TO_IGNORE` value, which reads the `pip freeze` internal API iviahttps://github.com/jazzband/pip-tools/blob/53309647980e2a4981db54c0033f98c61142de0b/piptools/_compat/pip_compat.py#L101

Simplest CI fix would likely be to drop those packages from the affected test cases, since their handling depends on the versions of other components.

**Comment by ncoghlan** [comment]
The `test_compile_recursive_extras_build_targets` failures feels like it might be due to https://github.com/pypa/pip/pull/12449 (`pip` used to inject `wheel` as an implicit build dependency for every `setup.py` project)

**Comment by chrysle** [comment]
> The `test_compile_recursive_extras_build_targets` failure feels like it might be due to [pypa/pip#12449](https://github.com/pypa/pip/pull/12449) (`pip` used to inject `wheel` as an implicit build dependency for every `setup.py` project)

Could you remove this line, then?

https://github.com/jazzband/pip-tools/blob/53309647980e2a4981db54c0033f98c61142de0b/tests/test_cli_compile.py#L3458

**Comment by chrysle** [comment]
> Some CI tests are failing due to this change in `pip` to no longer protect `setuptools`, `distribute` or `wheel` by default on Python 3.12+: [pypa/pip@5dc65ea](https://github.com/pypa/pip/commit/5dc65eabb75f89d4f4749b6c764042c227f6870a)

We don't test on 3.12+ yet, though...

**Comment by chrysle** [comment]
> The `test_compile_recursive_extras_build_targets` failure feels like it might be due to [pypa/pip#12449](https://github.com/pypa/pip/pull/12449) (`pip` used to inject `wheel` as an implicit build dependency for every `setup.py` project)

True, I've opened a PR, but there are still the [failing Windows tests](https://github.com/jazzband/pip-tools/actions/runs/9658693126/job/26640434682?pr=2106)... Could you look into them?

**Comment by ncoghlan** [comment]
> We don't test on 3.12+ yet, though...

I hit this one locally when trying to reproduce the other CI failures. I didn't check if the GitHub CI was running 3.12.

**Comment by webknjaz** [comment]
> The `test_compile_recursive_extras_build_targets` failure feels like it might be due to [pypa/pip#12449](https://github.com/pypa/pip/pull/12449) (`pip` used to inject `wheel` as an implicit build dependency for every `setup.py` project)

@ncoghlan no, apparently, this is because @agronholm moved `wheel` into `setuptools` during the sprints at PyCon US this spring. That failure actually led me to discovering a bug in the feature the failing test checks: https://github.com/jazzband/pip-tools/pull/1681#issuecomment-2212541289 / https://github.com/jazzband/pip-tools/pull/1681/files#r1667748889.

**Comment by ncoghlan** [comment]
Fixed the issue title to reflect that this isn't a CI issue (yet)

**Comment by ncoghlan** [comment]
From a windows testing point of view, I just ran `tox -e piplatest` locally, and the only reported failures were this (which I fixed by switching to my PR branch) and the `wheel` related failure that #2106 attempted to resolve.

**Comment by webknjaz** [comment]
> From a windows testing point of view, I just ran `tox -e piplatest` locally, and the only reported failures were this (which I fixed by switching to my PR branch) and the `wheel` related failure that #2106 attempted to resolve.

#2106 has been merged. We need to re-evaluate if this is still a problem.

**Comment by sirosen** [comment]
I'm reviewing old issues.

Pretty sure we got this in the course of #2188 -- I'm looking at potentially grabbing some changes out of #2105 , but this issue should be good to close.
